vscode/src/webviewPanel.ts

```diff
@@ -0,0 +1,240 @@
+            buckets.set(strKey, newValue);
+          }
+          const message = {
+            command: "update",
+            buckets: Array.from(buckets.entries()),
+            shotCount: resultCount,
+          };
+          sendMessageToPanel("histogram", false, message);
+        });
+
+        await worker.run(code, "", parseInt(numberOfShots), evtTarget);
+        clearTimeout(compilerTimeout);
+      } catch (e: any) {
+        log.error("Codegen error. ", e.toString());
+        throw new Error("Run failed");
+      } finally {
+        worker.terminate();
+      }
+    }),
+  );
+}
+
+type PanelType = "histogram" | "estimate" | "help";
+
+const panelTypeToPanel: Record<
+  PanelType,
+  { title: string; panel: QSharpWebViewPanel | undefined }
+> = {
+  histogram: { title: "Q# Histogram", panel: undefined },
+  estimate: { title: "Q# Estimates", panel: undefined },
+  help: { title: "Q# Help", panel: undefined },
+};
+
+function sendMessageToPanel(
+  panelType: PanelType,
+  reveal: boolean,
+  message: any,
+) {
+  const panelRecord = panelTypeToPanel[panelType];
+  if (!panelRecord.panel) {
+    const panel = window.createWebviewPanel(
+      panelType,
+      panelRecord.title,
+      ViewColumn.Beside,
+      {
+        enableScripts: true,
+        retainContextWhenHidden: true,
+      },
+    );
+
+    panelRecord.panel = new QSharpWebViewPanel(panelType, panel);
+  }
+
+  if (reveal) panelRecord.panel.reveal(ViewColumn.Beside);
+  if (message) panelRecord.panel.sendMessage(message);
+}
+
+export class QSharpWebViewPanel {
+  public static extensionUri: Uri;
+  private _ready = false;
+  private _queuedMessages: any[] = [];
+
+  constructor(
+    private type: PanelType,
+    private panel: WebviewPanel,
+  ) {
+    this.panel.onDidDispose(() => this.dispose());
+
+    this.panel.webview.html = this._getWebviewContent(this.panel.webview);
+    this._setWebviewMessageListener(this.panel.webview);
+  }
+
+  reveal(column: ViewColumn) {
+    this.panel.reveal(column);
+  }
+
+  private _getWebviewContent(webview: Webview) {
+    const extensionUri = QSharpWebViewPanel.extensionUri;
+
+    function getUri(pathList: string[]) {
+      return webview.asWebviewUri(Uri.joinPath(extensionUri, ...pathList));
+    }
+
+    const katexCss = getUri(["out", "katex", "katex.min.css"]);
+    const githubCss = getUri(["out", "katex", "github-markdown.css"]);
+    const webviewCss = getUri(["out", "webview", "webview.css"]);
+    const webviewJs = getUri(["out", "webview", "webview.js"]);
+
+    return /*html*/ `
+  <!DOCTYPE html>
+  <html lang="en">
+    <head>
+      <meta charset="UTF-8">
+      <meta name="viewport" content="width=device-width, initial-scale=1.0">
+      <title>Q#</title>
+      <link rel="stylesheet" href="${githubCss}" />
+      <link rel="stylesheet" href="${katexCss}" />
+      <link rel="stylesheet" href="${webviewCss}" />
+      <script src="${webviewJs}"></script>
```

Comment:
To simply the build and keep the number/size of scripts minimal, the WebView is actually identical html, js, and css for all scenarios, and renders different UX depending on the messages sent to it.